File: convert_articles.mjs
import { readFileSync, writeFileSync, mkdirSync, readdirSync, statSync } from 'fs';
import { join } from 'path';

const INPUT_FILE = '/Users/ming/hellosleep/docs/migrated-remote-articles-improved.json';
const OUTPUT_DIR = '/Users/ming/hellosleep/docs/articles';

function slugifyZh(text) {
  return text
    .replace(/[^\w\u4e00-\u9fff\u3400-\u4dbf\s-]/gu, '')
    .replace(/[\s_]+/g, '-')
    .replace(/^-+|-+$/g, '') || 'article';
}

function lexicalNodeToMd(node, depth = 0) {
  const type = node.type || '';
  const children = node.children || [];

  if (type === 'root') {
    return children.map(c => lexicalNodeToMd(c, depth)).filter(p => p.trim()).join('\n\n');
  }

  if (type === 'paragraph') {
    return children.map(c => lexicalNodeToMd(c, depth)).join('');
  }

  if (type === 'text') {
    let text = node.text || '';
    if (!text) return '';
    const fmt = node.format || 0;
    if (fmt & 16) text = `\`${text}\``;
    if (fmt & 1)  text = `**${text}**`;
    if (fmt & 2)  text = `*${text}*`;
    if (fmt & 4)  text = `~~${text}~~`;
    return text;
  }

  if (type === 'linebreak') return '\n';

  if (type === 'link') {
    const url = node.url || '';
    const linkText = children.map(c => lexicalNodeToMd(c, depth)).join('');
    return `[${linkText}](${url})`;
  }

  if (type === 'image') {
    const img = node.image || {};
    const url = img.url || '';
    const alt = img.alternativeText || img.name || 'image';
    const caption = img.caption || '';
    let result = `![${alt}](${url})`;
    if (caption) result += `\n*${caption}*`;
    return result;
  }

  if (type === 'heading') {
    const tag = node.tag || 'h2';
    const level = parseInt(tag[1]) || 2;
    const content = children.map(c => lexicalNodeToMd(c, depth)).join('');
    return `${'#'.repeat(level)} ${content}`;
  }

  if (type === 'quote') {
    const content = children.map(c => lexicalNodeToMd(c, depth)).join('');
    return content.split('\n').map(l => `> ${l}`).join('\n');
  }

  if (type === 'ul' || type === 'ol') {
    return children.map((child, i) => {
      const marker = type === 'ul' ? '-' : `${i + 1}.`;
      const indent = '  '.repeat(depth);
      return `${indent}${marker} ${lexicalNodeToMd(child, depth + 1)}`;
    }).join('\n');
  }

  if (type === 'listitem') {
    const textParts = [];
    const nested = [];
    for (const child of children) {
      if (child.type === 'ul' || child.type === 'ol') {
        nested.push('\n' + lexicalNodeToMd(child, depth + 1));
      } else {
        textParts.push(lexicalNodeToMd(child, depth));
      }
    }
    return textParts.join('') + nested.join('');
  }

  if (type === 'horizontalrule') return '---';

  if (type === 'code') {
    const lang = node.language || '';
    const lines = children.map(c =>
      c.type === 'linebreak' ? '\n' : lexicalNodeToMd(c, depth)
    );
    return `\`\`\`${lang}\n${lines.join('')}\n\`\`\``;
  }

  // Fallback
  return children.map(c => lexicalNodeToMd(c, depth)).join('');
}

function bodyToMarkdown(body) {
  if (typeof body === 'string') return body;
  if (Array.isArray(body)) {
    return body.map(node => lexicalNodeToMd(node)).filter(p => p.trim()).join('\n\n');
  }
  if (body && typeof body === 'object') return lexicalNodeToMd(body);
  return String(body);
}

function articleToMarkdown(article) {
  const data = article.data || article;
  const title = data.title || 'Untitled';
  const category = data.category || '';
  const date = data.date || '';
  const excerpt = data.excerpt || '';
  const body = data.body || [];

  const slug = slugifyZh(title);
  const filename = `${slug}.md`;

  const lines = [
    `# ${title}`,
    '',
    `**分类**: ${category}`,
    `**发布时间**: ${date}`,
    '',
  ];

  if (excerpt) {
    lines.push(`> ${excerpt}`, '');
  }

  lines.push(bodyToMarkdown(body));
  return { filename, content: lines.join('\n') };
}

// Main
mkdirSync(OUTPUT_DIR, { recursive: true });

const articles = JSON.parse(readFileSync(INPUT_FILE, 'utf-8'));
console.log(`Total articles: ${articles.length}`);

const slugCounts = {};
let written = 0;
let skipped = 0;

for (const article of articles) {
  try {
    let { filename, content } = articleToMarkdown(article);
    const base = filename.slice(0, -3);
    if (base in slugCounts) {
      slugCounts[base]++;
      filename = `${base}-${slugCounts[base]}.md`;
    } else {
      slugCounts[base] = 0;
    }

    writeFileSync(join(OUTPUT_DIR, filename), content, 'utf-8');
    written++;
  } catch (e) {
    const title = (article.data || article).title || '?';
    console.error(`  ERROR [${title}]: ${e.message}`);
    skipped++;
  }
}

console.log(`\nDone! Written: ${written}, Skipped: ${skipped}`);

const files = readdirSync(OUTPUT_DIR).sort();
console.log(`\nFirst 10 of ${files.length} files:`);
files.slice(0, 10).forEach(name => {
  const size = statSync(join(OUTPUT_DIR, name)).size;
  console.log(`  ${name}  (${size} bytes)`);
});

File: convert_articles.py
#!/usr/bin/env python3
"""Convert migrated-remote-articles-improved.json to individual Markdown files."""

import json
import os
import re
import unicodedata

INPUT_FILE = "/Users/ming/hellosleep/docs/migrated-remote-articles-improved.json"
OUTPUT_DIR = "/Users/ming/hellosleep/docs/articles"


def slugify_zh(text: str) -> str:
    """Slugify preserving Chinese characters."""
    text = re.sub(r"[^\w\u4e00-\u9fff\u3400-\u4dbf\s-]", "", text, flags=re.UNICODE)
    text = re.sub(r"[\s_]+", "-", text)
    return text.strip("-") or "article"


def lexical_node_to_md(node: dict, depth: int = 0) -> str:
    """Recursively convert a Lexical JSON node to Markdown."""
    node_type = node.get("type", "")
    children = node.get("children", [])

    if node_type == "root":
        parts = [lexical_node_to_md(c, depth) for c in children]
        return "\n\n".join(p for p in parts if p.strip())

    if node_type == "paragraph":
        return "".join(lexical_node_to_md(c, depth) for c in children)

    if node_type == "text":
        text = node.get("text", "")
        fmt = node.get("format", 0)
        if not text:
            return ""
        if fmt & 16:
            text = f"`{text}`"
        if fmt & 1:
            text = f"**{text}**"
        if fmt & 2:
            text = f"*{text}*"
        if fmt & 4:
            text = f"~~{text}~~"
        return text

    if node_type == "linebreak":
        return "\n"

    if node_type == "link":
        url = node.get("url", "")
        link_text = "".join(lexical_node_to_md(c, depth) for c in children)
        return f"[{link_text}]({url})"

    if node_type == "image":
        img = node.get("image", {})
        url = img.get("url", "")
        alt = img.get("alternativeText", "") or img.get("name", "image")
        caption = img.get("caption", "")
        result = f"![{alt}]({url})"
        if caption:
            result += f"\n*{caption}*"
        return result

    if node_type == "heading":
        tag = node.get("tag", "h2")
        level = int(tag[1]) if tag and len(tag) == 2 and tag[1].isdigit() else 2
        content = "".join(lexical_node_to_md(c, depth) for c in children)
        return f"{'#' * level} {content}"

    if node_type == "quote":
        content = "".join(lexical_node_to_md(c, depth) for c in children)
        return "\n".join(f"> {line}" for line in content.split("\n"))

    if node_type in ("ul", "ol"):
        items = []
        for i, child in enumerate(children):
            marker = "-" if node_type == "ul" else f"{i + 1}."
            indent = "  " * depth
            items.append(f"{indent}{marker} {lexical_node_to_md(child, depth + 1)}")
        return "\n".join(items)

    if node_type == "listitem":
        text_parts = []
        nested = []
        for child in children:
            if child.get("type") in ("ul", "ol"):
                nested.append("\n" + lexical_node_to_md(child, depth + 1))
            else:
                text_parts.append(lexical_node_to_md(child, depth))
        return "".join(text_parts) + "".join(nested)

    if node_type == "horizontalrule":
        return "---"

    if node_type == "code":
        lang = node.get("language", "")
        lines = []
        for child in children:
            if child.get("type") == "linebreak":
                lines.append("\n")
            else:
                lines.append(lexical_node_to_md(child, depth))
        return f"```{lang}\n{''.join(lines)}\n```"

    # Fallback
    return "".join(lexical_node_to_md(c, depth) for c in children)


def body_to_markdown(body) -> str:
    if isinstance(body, str):
        return body
    if isinstance(body, list):
        parts = [lexical_node_to_md(node) for node in body]
        return "\n\n".join(p for p in parts if p.strip())
    if isinstance(body, dict):
        return lexical_node_to_md(body)
    return str(body)


def article_to_markdown(article: dict) -> tuple:
    data = article.get("data", article)
    title = data.get("title", "Untitled")
    category = data.get("category", "")
    date = data.get("date", "")
    excerpt = data.get("excerpt", "")
    body = data.get("body", [])

    slug = slugify_zh(title)
    filename = f"{slug}.md"

    lines = [
        f"# {title}",
        "",
        f"**分类**: {category}",
        f"**发布时间**: {date}",
        "",
    ]
    if excerpt:
        lines += [f"> {excerpt}", ""]

    lines.append(body_to_markdown(body))
    return filename, "\n".join(lines)


def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    with open(INPUT_FILE, encoding="utf-8") as f:
        articles = json.load(f)

    print(f"Total articles: {len(articles)}")

    slug_counts = {}
    written = 0
    skipped = 0

    for article in articles:
        try:
            filename, content = article_to_markdown(article)
            base = filename[:-3]
            if base in slug_counts:
                slug_counts[base] += 1
                filename = f"{base}-{slug_counts[base]}.md"
            else:
                slug_counts[base] = 0

            path = os.path.join(OUTPUT_DIR, filename)
            with open(path, "w", encoding="utf-8") as f:
                f.write(content)
            written += 1
        except Exception as e:
            title = article.get("data", article).get("title", "?")
            print(f"  ERROR [{title}]: {e}")
            skipped += 1

    print(f"\nDone! Written: {written}, Skipped: {skipped}")
    files = sorted(os.listdir(OUTPUT_DIR))
    print(f"\nFirst 10 of {len(files)} files:")
    for name in files[:10]:
        size = os.path.getsize(os.path.join(OUTPUT_DIR, name))
        print(f"  {name}  ({size} bytes)")


if __name__ == "__main__":
    main()
